### PR TITLE
Native hosting: validate frameworks and properties when initializing secondary context

### PR DESCRIPTION
**Pull request by elinor-fung**
- Fail initialization if the frameworks referenced in the config for a secondary context do not exist or are incompatible with the frameworks used for the first (active) context
- Return specific status code if runtime properties specified in the config for a secondary context do not exist or have different values from the properties used to load the runtime
- Add tests for compatibility check of secondary context when first context was created using non-context-based APIs and context-based APIs (and collapse some code duplication)

### CODE REVIEW

File: src/corehost/cli/fxr/fx_muxer.cpp [hunk 1]
```diff
@@ -839,7 +840,9 @@ namespace
             return StatusCode::InvalidConfigFile;
         }
 
-        // [TODO] Validate the current context is acceptable for this request (frameworks)
+        // Validate the current context is acceptable for this request (frameworks)
+        if (!fx_resolver_t::is_config_compatible_with_frameworks(app_config, existing_context->fx_versions_by_name))
+            return StatusCode::CoreHostIncompatibleConfig;
```
Comment (sdmaclea) [inline src/corehost/cli/fxr/fx_muxer.cpp]: Do we always want a hard fail here?

File: src/corehost/cli/hostpolicy/hostpolicy.cpp [hunk 2]
```diff
@@ -636,7 +670,13 @@ SHARED_API int __cdecl corehost_initialize(const corehost_initialize_request_t *
...
     if (rc == StatusCode::CoreHostAlreadyInitialized)
     {
-        // [TODO] Compare the current context with this request (properties)
+        assert(init_request != nullptr
+            && init_request->version >= offsetof(corehost_initialize_request_t, config_values) + sizeof(init_request->config_values)
+            && init_request->config_keys.len == init_request->config_values.len);
+
+        // Compare the current context with this request (properties)
+        if (!matches_existing_properties(init_request))
+            rc = StatusCode::CoreHostDifferentProperties;
```
Comment (sdmaclea) [inline src/corehost/cli/hostpolicy/hostpolicy.cpp]: This seems strange.  I thought we were going to let the second client observe the properties and make the decision for themselves. This seems to be forcing failure.

File: src/corehost/cli/fxr/fx_muxer.cpp (hunk 1 above)
Comment (vitek-karas) [reply src/corehost/cli/fxr/fx_muxer.cpp]: At least for now I think yes. I could imagine that in the future we allow the native app to override this decision, but for now I think it's safer to not go there yet.

File: src/corehost/cli/hostpolicy/hostpolicy.cpp (hunk 2 above)
Comment (vitek-karas) [reply src/corehost/cli/hostpolicy/hostpolicy.cpp]: `StatusCode::CoreHostDifferentProperties` is a "warning" error code. It will be documented that it's perfectly valid to use the returned context.

File: src/test/HostActivationTests/NativeHosting/HostContext.cs [hunk 3]
```diff
@@ -224,166 +160,368 @@ public void GetDelegate_Multiple(string checkProperties)
...
+            propertyValidation.ValidateSecondaryContext(result, SharedTestState.ConfigPropertyName, SharedTestState.ConfigPropertyValue);
+        }
+
+        [Theory]
+        [InlineData(Scenario.ConfigMultiple,  Constants.MicrosoftNETCoreApp, "1.1.0", Constants.RollForwardSetting.LatestMinor, false)]
+        [InlineData(Scenario.ConfigMultiple,  Constants.MicrosoftNETCoreApp, "1.1.0", Constants.RollForwardSetting.LatestMajor, true)]
+        [InlineData(Scenario.ConfigMultiple,  Constants.MicrosoftNETCoreApp, "2.1.0", Constants.RollForwardSetting.LatestPatch, false)]
+        [InlineData(Scenario.ConfigMultiple,  Constants.MicrosoftNETCoreApp, "2.1.0", Constants.RollForwardSetting.LatestMinor, true)]
+        [InlineData(Scenario.ConfigMultiple,  Constants.MicrosoftNETCoreApp, "2.1.0", Constants.RollForwardSetting.LatestMajor, true)]
+        [InlineData(Scenario.ConfigMultiple,  Constants.MicrosoftNETCoreApp, "3.1.0", Constants.RollForwardSetting.LatestMinor, false)]
```
Comment (vitek-karas) [inline src/test/HostActivationTests/NativeHosting/HostContext.cs]: Can you please add at least for one of the scenarios also the other roll-forward values. I know that components should use the LatestMinor or LatestMajor most of the time, but the others should still work.

File: src/test/HostActivationTests/NativeHosting/HostContext.cs [hunk 4]
```diff
@@ -224,166 +160,368 @@ public void GetDelegate_Multiple(string checkProperties)
...
                     break;
-                default:
-                    throw new Exception($"Unknown option: {checkProperties}");
+            }
+
+            if (existingValue == null)
+            {
+                result.Should()
+                    .InitializeSecondaryContext(propertyCompatConfig, CoreHostDifferentProperties)
+                    .And.HaveStdErrContaining($"The property [{propertyName}] is not present in the previously loaded runtime");
```
Comment (vitek-karas) [inline src/test/HostActivationTests/NativeHosting/HostContext.cs]: Can we also validate the return code from the call? This should return the special code...

File: src/test/HostActivationTests/NativeHosting/HostContext.cs [hunk 5]
```diff
@@ -224,166 +160,368 @@ public void GetDelegate_Multiple(string checkProperties)
...
+                case Scenario.Mixed:
                     result.Should()
-                        .FailToGetRuntimePropertyValue(LogPrefix.App, sharedState.AppPropertyName, HostInvalidState)
-                        .And.FailToGetRuntimePropertyValue(LogPrefix.App, sharedState.ConfigPropertyName, HostInvalidState)
-                        .And.GetRuntimePropertyValue(LogPrefix.Secondary, sharedState.AppPropertyName, sharedState.AppPropertyValue)
-                        .And.FailToGetRuntimePropertyValue(LogPrefix.Secondary, sharedState.ConfigPropertyName, HostPropertyNotFound)
-                        .And.HavePropertyMock(sharedState.AppPropertyName, sharedState.AppPropertyValue)
-                        .And.NotHavePropertyMock(sharedState.ConfigPropertyName);
+                        .InitializeContextForApp(appOrConfigPath)
+                        .And.ExecuteAssemblyMock(appOrConfigPath, new string[0]);
```
Comment (vitek-karas) [inline src/test/HostActivationTests/NativeHosting/HostContext.cs]: This should also create the delegate mock, right?

File: src/test/HostActivationTests/NativeHosting/HostContext.cs [hunk 6]
```diff
@@ -224,166 +160,368 @@ public void GetDelegate_Multiple(string checkProperties)
...
-                case CheckProperties.GetAllActive:
+                case Scenario.NonContextMixed:
                     result.Should()
-                        .FailToGetRuntimeProperties(LogPrefix.App, HostInvalidState)
-                        .And.GetRuntimePropertiesIncludes(LogPrefix.Secondary, sharedState.AppPropertyName, sharedState.AppPropertyValue)
-                        .And.GetRuntimePropertiesExcludes(LogPrefix.Secondary, sharedState.ConfigPropertyName)
-                        .And.HavePropertyMock(sharedState.AppPropertyName, sharedState.AppPropertyValue)
-                        .And.NotHavePropertyMock(sharedState.ConfigPropertyName);
+                        .ExecuteAssemblyMock(appOrConfigPath, new string[0]);
                     break;
```
Comment (vitek-karas) [inline src/test/HostActivationTests/NativeHosting/HostContext.cs]: This case should also create a delegeate mock...

Comment (vitek-karas) [reply src/test/HostActivationTests/NativeHosting/HostContext.cs]: Actually all of these cases should also create the second context delegate mock... ideally we would ask for a different delegate and verify that it got created...
This also applies to the framework test above (but inly in the success cases)

File: src/test/HostActivationTests/NativeHosting/HostContext.cs [hunk 7]
```diff
@@ -224,166 +160,368 @@ public void GetDelegate_Multiple(string checkProperties)
...
+                result.Should().Fail()
+                    .And.FailToInitializeContextForConfig(CoreHostIncompatibleConfig)
+                    .And.HaveStdErrContaining($"The specified framework '{frameworkName}' is not present in the previously loaded runtime");
+            }
+        }
+
+        [Theory]
+        [InlineData(Scenario.ConfigMultiple,  SharedTestState.ConfigPropertyName, SharedTestState.ConfigPropertyValue,   SharedTestState.ConfigPropertyValue)]
+        [InlineData(Scenario.ConfigMultiple,  SharedTestState.ConfigPropertyName, "NEW_PROPERTY_VALUE",                  SharedTestState.ConfigPropertyValue)]
+        [InlineData(Scenario.ConfigMultiple,  "NEW_PROPERTY_NAME",                "NEW_PROPERTY_VALUE",                  null)]
```
Comment (vitek-karas) [inline src/test/HostActivationTests/NativeHosting/HostContext.cs]: Please add a test with empty property values and also another one where the property value difference is small:
* One is a substring of the other
* They only differ in casing

Ideally also a test with multiple properties in both configs..., but that is probably trickier to setup.

File: src/corehost/cli/hostpolicy/hostpolicy.cpp (hunk 2 above)
Comment (elinor-fung) [reply src/corehost/cli/hostpolicy/hostpolicy.cpp]: It is intended to still be considered success and not failure (just like `StatusCode::CoreHostAlreadyInitialized`). I basically stole the COM approach and made > 0 values different success codes.

Would it be clearer to give the success codes some prefix (`Success*`? `Warning*`?) indicating they are not actually failures?

Comment (vitek-karas) [reply src/corehost/cli/hostpolicy/hostpolicy.cpp]: That's a good idea...
`Success_DifferentRuntimeProperties` - would make lot of sense to me.

File: src/test/HostActivationTests/NativeHosting/HostContext.cs (hunk 4 above)
Comment (elinor-fung) [reply src/test/HostActivationTests/NativeHosting/HostContext.cs]: `InitializeSecondaryContext` checks the return code (that the nativehost test exe prints)?

File: src/test/HostActivationTests/NativeHosting/HostContext.cs (hunk 7 above)
Comment (elinor-fung) [reply src/test/HostActivationTests/NativeHosting/HostContext.cs]: Made it so that there are cases with multiple properties. The set up I ended up with ('multiple' is really just two) is a little awkward though.

File: src/test/HostActivationTests/NativeHosting/HostContext.cs (hunk 6 above)
Comment (elinor-fung) [reply src/test/HostActivationTests/NativeHosting/HostContext.cs]: The native host is now always using the `com_activation` type when getting a delegate with the active context and `load_in_memory_assembly` with the secondary context (`CreateDelegateMock_COM/InMemoryAssembly` check the args the mock coreclr prints).
